_posts/pl/2024-07-26-czy-wiesz-jak-obslugiwac-bledy-w-rxjs.md

```diff
@@ -0,0 +1,86 @@
+---
+layout:    post
+title:     Czy wiesz, jak obsługiwać błedy w rxjs?
+date:      2024-06-26T08:00:00+01:00
+published: true
+didyouknow: false
+lang: pl
+author: ptatarski
+image: /assets/img/posts/2024-07-26-czy-wiesz-jak-obslugiwac-bledy-w-rxjs/thumbnail.webp
+tags:
+- rxjs
+- javascript
+---
+
+Biblioteka rxjs dostarcza nam kilka mechanizmów, które ułatwiają nam reagowanie na nieprzewidziane sytuacje występujące podczas procesowania strumienia danych w naszej aplikacji.
+
+## Operator catchError
+
+Jednym z najpopularniejszych jest operator `catchError`. Operator ten pozwala nam zareagować na sytuację, kiedy w strumieniu z jakiegoś powodu nagle wystąpi błąd. Zamiast brzydkiego błędu w konsoli możemy w `catchError` dostarczyć `Observable`, który będzie przetwarzany dalej w strumieniu.
+
+Ważne, aby `catchError` znajdował się po operacji, w której wystąpi błąd, inaczej ten nie zostanie przechwycony.
+
+Przykładowy kod:
+```javascript
+of(1, 2, 3, 4, 5)
+  .pipe(
+    catchError(() => of(99)), // Ten catchError nie obsłuży błędu występującego niżej
+    map((num) => {
+      if (num === 4) {
+        throw new Error('test');
+      }
+      return num;
+    }),
+    catchError((error) => of(`Złapano błąd ${error}`))
+  )
+  .subscribe({
+    next: console.log,
+    error: (err) => console.log('error', err.message),
+    complete: () => console.log('completed'),
+  });
+// 1
+// 2
+// 3
+// Złapano błąd Error: test
+// completed
+```
+Jak widać błąd został złapany i nigdy nie wpadł w obsługę `error observera`.
+
+## Operator retry
+
+Kolejnym operatorem pozwalającym na obsługę błędów jest operator `retry`. Jak sama nazwa mówi operator ten pozwala na ponowienie operacji, co jest przydatne jeżeli zakładamy, że operacja w strumieniu może się zakończyć niepowodzeniem z przyczyn niezależnych od użytkownika np. niedostępności usługi. `Retry` ponowi wtedy obsługę zaczynając od początku strumienia.
```

Comment:
Jak sama nazwa wskazuje, operator ten pozwala na ponowienie operacji. Jest to przydatne, jeżeli zakładamy, że operacja w strumieniu może się zakończyć niepowodzeniem z przyczyn niezależnych od użytkownika, np. niedostępności usługi.

`Retry` ponowi wtedy obsługę, zaczynając od początku strumienia.